docs/dev-notes/BACKGROUND-QUEUE.md

```diff
@@ -0,0 +1,90 @@
+# Background queue 
+
+## What is the background queue? 
+
+Anytime that the SDK needs to perform any asynchronous operation (such as a HTTP request), the SDK will simply add the task to a queue and run the task at a later time instead of running the operation right away. 
+
+By running tasks at a later time, there are many great benefits. 
+1. The SDK performs error handling and retries tasks later instead of the customer needing to do it. 
+2. The SDK is much more powerful now that we can add any task to the queue
+3. We can provide offline functionality to our SDK. 
+4. We save battery on the device because we batch network requests together. 
+5. We can provide a better design of the SDK functions for customers. 
+
+Instead of designing our SDK to be asynchronous like this:
+
+```swift 
+CustomerIO.shared.identify("Dana") { result in 
+    switch result {
+        case .success: 
+            // yay! Profile identified! 
+        case .failure(let cioError): 
+            // Handle failure. Determine what failed and why. 
+            // Write custom code to maybe retry the request later. 
+    }   
+}
+```
+
+With a queue, we can design our SDK to be similar to this:
+
+```swift 
+CustomerIO.shared.identify("Dana")
+```
+
+## How do I QA test the background queue? 
+
+See [the QA doc](QA.md) in the background queue section. 
+
+## How is the background queue built? 
+
+The queue has a few main concepts. 
+
+* Adding a task to the queue. 
+
+Tasks need to be saved to persistent storage on the device to be able to run the task later on. We save .json files on the file system for the queue. We choose JSON simply because it's an easy way to save/read Swift objects to files. 
+
+To be memory efficient, the queue data structure (aka: queue inventory) is saved in 1 JSON file. This inventory is general metadata about the tasks in the queue (see file `QueueTaskMetadata`). Then, each task of the queue is stored in it's own individual JSON file. 
+
+Here is an idea of the file system structure where the queue has 2 tasks in it:
+
+```
+Documents/
+  io.customer.ios/
+    queue/
+      inventory.json
+      tasks/
+        3838383949849493939393.json
+        2939929292001919202002.json
```

Comment:
Each task will have a separate JSON file ? Suppose out of 10 tasks , 2 tasks didn't execute successfully. Files related to those tasks will be deleted (and then recreated) or they will be there in the local file structure and removed only when tasks are done ?

Reply:
Good question. I added a note about this to the doc. 

> After a task in the queue executes successfully, the queue will update the `inventory.json` file and will delete the task's `.json` file in `tasks/`. See `QueueStorage.swift` to view all the file system operations performed by the queue.

Reply:
Great. Thanks. 👍